web/modules/custom/joinup_core/src/GroupHelper.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_core;
+
+use Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException;
+use Drupal\Component\Plugin\Exception\PluginNotFoundException;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\rdf_entity\Entity\Rdf;
+use Drupal\rdf_entity\RdfEntitySparqlStorageInterface;
+use Drupal\rdf_entity\RdfInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Helper methods for working with groups (collections and solutions).
+ */
+class GroupHelper implements GroupHelperInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * A logger instance.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * Constructs a new GroupHelper.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   A logger instance.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, LoggerInterface $logger) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->logger = $logger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getGroupByLabel(string $label): RdfInterface {
+    $query = $this->getRdfEntityStorage()->getQuery()
+      ->condition('label', $label)
+      ->condition('rid', self::BUNDLES, 'IN');
+    $result = $query->execute();
+
+    if (empty($result)) {
+      throw new \InvalidArgumentException("No group found with label '$label'.");
+    }
+
+    if (count($result) > 1) {
+      $this->logger->warning("Found multiple groups with the label '$label'.");
+    }
+
+    return Rdf::load(reset($result));
+  }
+
+  /**
+   * The storage handler for RDF Entities.
+   *
+   * @return \Drupal\rdf_entity\RdfEntitySparqlStorageInterface
+   *   The storage handler.
+   */
+  protected function getRdfEntityStorage(): RdfEntitySparqlStorageInterface {
+    try {
+      return $this->entityTypeManager->getStorage('rdf_entity');
+    }
+    catch (PluginNotFoundException $e) {
```

Comment:
No it is a different situation, the `$this->entityTypeManager` is known to be not NULL since the class cannot be instantiated otherwise.

You are misunderstanding the comment. The comment is on the line about throwing the `RuntimeException`. I am explaining _why_ this is being thrown. I am not explaining why this is being caught, because I think it is clear and doesn't need explanation. We need to catch these exceptions since they are defined as checked exceptions on the interface. Ref `EntityTypeManagerInterface::getStorage()`:

```
  /**
   * Creates a new storage instance.
   *
   * @param string $entity_type
   *   The entity type for this storage.
   *
   * @return \Drupal\Core\Entity\EntityStorageInterface
   *   A storage instance.
   *
   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
   *   Thrown if the entity type doesn't exist.
   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
   *   Thrown if the storage handler couldn't be loaded.
   */
  public function getStorage($entity_type);
```

We need to either catch them or document them if it is the responsibility of the calling code to catch it. It is not however in this case, it is our responsibility, we are instantiating the EntityTypeManager so we need to deal with their exceptions.

Would you prefer this to be logged rather than rethrown? I think it is much more convenient to rethrow if the exception "should not happen". It think it warrants the extra work to inject the logger for this case.

> We are going to end up writing 17 lines to do something for a single line every time?

Well yes we need to catch these exceptions as is demanded by core. But we can provide a trait if you like.

Reply:
Thinking a bit more about this, a trait would not be the right solution. I think the right way according to the architecture would be to wrap the entire `EntityManagerInterface` in a specific `RdfEntityManagerInterface` which would then omit the `$entity_type` arguments and narrow down all the return types to Rdf specific types.

In the implementation of this `RdfEntityManager` we can then catch all the generic exceptions and convert them to runtime exceptions, and then the calling code can completely rely on the correct types being returned, cleaning up the code immensely.

I will discuss this with @sandervd to get his opinion.

Reply:
I didn't get an answer yet from Sander but there are a lot of methods on the `EntityManager` so it would be too much work for this PR.

This practice is still quite uncommon in PHP so I am not suggesting that everyone should adopt it right now and we should convert our entire code base. I think we will see it gaining popularity now that it is detected in PHPStorm. At least I started using it :) I would argue to not force to adopt it, but also not remove it when people want to do the additional effort to keep our code clean.

Reply:
I am not arguing against the idea of having an exception caught as a RuntimeException or even catching it at all. I am arguing over the cost of readability and polluting the code with unecessary catches only to catch an exception that will be anyway thrown but as a runtime instead of a PluginNotFoundException.
My argument is that this will happen once or twice in the lifetime of the project in cases like this, even more because we are requesting the `rdf_entity` storage. If this is not found, there is something so wrong with the codebase that it is not a RuntimeException.. It is a storage failure or something.
Furthermore, the fact that we are requesting a hardcoded storage, means that this is meant to work. It is not like we are looping over an arbitrary array of strings and we try to get the storage.

Anyway, I cannot argue against you guys enough. I cannot reject this because there is no end to this discussion.. 
If we were to accept that, we would have to also catch the `\Drupal\Core\Entity\EntityStorageInterface::save` `EntityStorageException` and then every other case that PHPStorm reports. I know that this is not an issue with Drupal but with the database, but still it can be considered a RuntimeException as well.
I don't think it will gain any popularity and even if it does, it will not be left to each project to catch every case. It might be dealt in core.

Reply:
Yes the `EntityStorageException` is also a checked exception so it should be either caught or documented on the interface if the calling code should deal with it.

This is not about coding standards or PHPStorm inspections but about correctly implementing the interfaces that we are using. PHPStorm is simply reading the interfaces and alerting us if we did not implement it correctly.

Reply:
This code is removed now, so the problem is gone :)